src/levanter/generator.py

```diff
@@ -0,0 +1,296 @@
+import time
+from dataclasses import dataclass
+from typing import Any, Callable, Literal, Tuple
+
+import equinox as eqx
+import jax
+import numpy as np
+from jax.random import PRNGKey
+from jaxtyping import PRNGKeyArray
+from tqdm.auto import tqdm
+from typing_extensions import Self
+
+from haliax import Axis, NamedArray
+from haliax.nn import hax
+
+from levanter.compat.hf_checkpoints import HFCompatConfig
+from levanter.models.lm_model import LmConfig, LmHeadModel
+from levanter.utils.hf_utils import HfTokenizer
+
+
+Batch = hax.Axis("batch", 1)  # TODO: Verify multi-batch support
+Config = LmConfig | HFCompatConfig
+
+
+@dataclass
+class SamplingOutput:
+    """Struct for encapsulating the model output in a convenience wrapper."""
+
+    output_tokens: list[int]
+    num_tokens: int
+    decoded_output: str
+    total_time_taken: float
+
+    def __repr__(self: Self) -> str:
+        return f"\nFull decoded sentence:\n{'=' * 10}\n{self.decoded_output}\n{'=' * 10}"
+
+
+@dataclass(frozen=True)
+class Prompt:
+    """Struct for encapsulating the prompt and pretty print it."""
+
+    input_prompt: str
+    system_instruction: str = ""
+    user_format: str = ""
+
+    def _to_string(self: Self) -> str:
+        return (
+            "Below is an instruction that describes a task, paired with an input that provides further context. Write"
+            " a response that appropriately completes the request in the provided format.\n\n### Instruction:\n"
+            f" {self.system_instruction}\n### Input:\n {self.input_prompt}\n### Format: {self.user_format}###"
+            " Response:"
+        )
+
+    def __repr__(self: Self) -> str:
+        return f"{'-' * 10}\nModel Prompt:{self._to_string()}\n{'-' * 10}"
+
+
+@dataclass(frozen=True)
+class SamplingArgs:
+    """Parent struct for encapsulating arguments provided to a sampler."""
+
+    prompt: Prompt
+    num_tokens: int = 64
+    temperature: float = 0.69
+
+
+@dataclass(frozen=True, kw_only=True)
+class GreedySamplerArgs(SamplingArgs):
+    """Struct to encapsulate arguments for the Greedy Sampler."""
+
+    repetition_token_window: int = 4
+    repetition_penalty: float = 0.8
+    top_k: int = 50
+    top_p: float = 0.8
+
+
+class Generator:
+    hf_config: Config
```

Comment:
our main use case is going to be evals during training, so it would be better if it took an LmHeadModel instead (or a subclass that supported efficient inference or something)